tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogicTest.java

```diff
@@ -474,6 +475,23 @@ public void testWholeWordSearchAvailable() {
 		assertThat(findReplaceLogic.isWholeWordSearchAvailable(""), is(false));
 	}
 
+	public void testReplaceInSelectionStaysInSelection() {
+		TextViewer textViewer= setupTextViewer("line" + System.lineSeparator() + "line" + System.lineSeparator() + "line");
+		textViewer.setSelectedRange(5, 9);
+		IFindReplaceLogic findReplaceLogic= setupFindReplaceLogicObject(textViewer);
+		findReplaceLogic.deactivate(SearchOptions.GLOBAL);
+		findReplaceLogic.performSelectAndReplace("line", "");
+		expectStatusEmpty(findReplaceLogic);
+
+		findReplaceLogic.performSelectAndReplace("line", "");
+		assertThat(textViewer.getTextWidget().getText(), is("line" + System.lineSeparator() + System.lineSeparator()));
+		expectStatusEmpty(findReplaceLogic);
```

Comment:
Can you explain this behavior, @Wittmaxi?
The test initially restricts the find/replace application to a selection containing the second line of the document. Then you replace the second line with empty contents and then assert that doing another replace call also replaces the third line. But actually the third line is not part of the initial selection. So I would expect that the second replace call does nothing. Why is the defined behavior expected/valid?

Reply:
I am not sure if I understood you correctly, I will think out loud - please tell me if that addresses your question;

The function-signature of setSelectedRange is ```public void setSelectedRange(int selectionOffset, int selectionLength)```, so I set a selection that begins at the 5th Character and is nine character long. that could look like this:
```
line\n[line\nline] on linux
line\r[\nline\r\nli]ne on windows
```

the length of the selection is 4 + 4 + 1 or 4 + 4 + 2 on windows

Keep in mind that it is not crucial that the ENTIRE line is selected (hence why we can tolerate variable lineSeparator-width): a search scope will automatically expand to be a whole line if only a part of it is selected, as seen here:
![180420221002](https://github.com/eclipse-platform/eclipse.platform.ui/assets/16443184/44bfab1f-cf54-4b80-a15f-b31bbf1e85ad)

With this selection in mind, performing replace once expects to get
```
line\r\n\r\nline
```
and on the second replace, I expect to get
```
line\r\n\r\n
```

On the third replace, I don't expect any changes (which is what this test covers!)

Reply:
Sorry for the confusion, I took the "9" to be the end index and not the length of the selection.

Still, the test fails for me on the first assertion when activated (note that it currently misses an `@Test` annotation and is  thus not executed at all).

Reply:
Indeed, I did not see that the test wasn't activated and assumed the test ran correctly - hence our misunderstanding.
The test fails becaue I did not include the fix to this unit-test into this PR which I had added to the Overlay-PR. I have fixed this, at least on my computer the (now activated) test runs fine.

Reply:
I see. Can you explain why you have removed the assertion after the first replacement?
```java
assertThat(textViewer.getTextWidget().getText(), is("line" + System.lineSeparator() + System.lineSeparator() + "line"));
```
If I readd that assertion, the test fails. The failure shows that the first `performSelectAndReplace()` call replaces the third line instead of the second:
```
Expected: is "line\r\n\r\nline"
     but: was "line\r\nline\r\n"
```

Isn't that a bug?

Reply:
It seems to me that this is indeed a bug. I have traced it down to being caused by a "perform search" in a scope not working properly. Here is a unit-test which reproduces that problem:

```java

	@Test
	public void testFirstSearchInScope() {
		String lineSeparator= System.lineSeparator();
		TextViewer textViewer= setupTextViewer("line" + lineSeparator + "line" + lineSeparator + "line");
		int lineSeparatorLength= lineSeparator.length();
		textViewer.setSelectedRange(4 + lineSeparatorLength, 8 + lineSeparatorLength);
		IFindReplaceLogic findReplaceLogic= setupFindReplaceLogicObject(textViewer);
		findReplaceLogic.performSearch("line");
		expectStatusEmpty(findReplaceLogic);
		assertThat(findReplaceLogic.getTarget().getSelection().x, not(is(0)));
}
```

This behavior might be cause because the new search-offset isn't set when I update the searchscope.
In the "real-world", the bug doesn't seem to show:

![1](https://github.com/eclipse-platform/eclipse.platform.ui/assets/16443184/d88ed9ae-6f40-4195-a12c-f2d02d8bf3b5)

Reply:
The unit Test I provided for reproducing this bug is missing activation of the search-scope.
This is the new test:

```java
	@Test
	public void testFirstSearchInScope() {
		String lineSeparator= System.lineSeparator();
		TextViewer textViewer= setupTextViewer("line" + lineSeparator + "line" + lineSeparator + "line");
		int lineSeparatorLength= lineSeparator.length();
		textViewer.setSelectedRange(4 + lineSeparatorLength, 8 + lineSeparatorLength);
		IFindReplaceLogic findReplaceLogic= setupFindReplaceLogicObject(textViewer);
		findReplaceLogic.activate(SearchOptions.GLOBAL);
		findReplaceLogic.performSearch("line");
		expectStatusEmpty(findReplaceLogic);
		assertThat(findReplaceLogic.getTarget().getSelection().x, not(is(0)));
	}
```

It doesn't fail, so it doesn't reproduce what is happening. I will keep the Unit-test anyway because it tests something important and untested so far

Reply:
```java
	@Test
	public void testFirstSearchInScope() {
		String lineSeparator= System.lineSeparator();
		TextViewer textViewer= setupTextViewer("line" + lineSeparator + "line" + lineSeparator + "line");
		int lineSeparatorLength= lineSeparator.length();
		textViewer.setSelectedRange(4 + lineSeparatorLength, 8 + lineSeparatorLength);
		IFindReplaceLogic findReplaceLogic= setupFindReplaceLogicObject(textViewer);
		findReplaceLogic.deactivate(SearchOptions.GLOBAL);
		findReplaceLogic.performSearch("line");
		expectStatusEmpty(findReplaceLogic);
		assertThat(findReplaceLogic.getTarget().getSelection().x, not(is(0)));
		assertThat(findReplaceLogic.getTarget().getSelection().x, not(is(textViewer.getDocument().get().length() - 4)));
	}
```

This is the test we are looking for

Reply:
I have found the cause of this strange bug:

When initializing the searchScope, the new "search-cursor" is set, like this:

```java
	/**
	 * Tells the dialog to perform searches only in the scope given by the actually
	 * selected lines.
	 */
	private void initializeSearchScope() {
		if (shouldInitIncrementalBaseLocation()) {
			initIncrementalBaseLocation();
		}

		if (target == null || !(target instanceof IFindReplaceTargetExtension)) {
			return;
		}

		IFindReplaceTargetExtension extensionTarget = (IFindReplaceTargetExtension) target;
		IRegion scope;
		Point lineSelection = extensionTarget.getLineSelection();

		scope = new Region(lineSelection.x, lineSelection.y);
		extensionTarget.setScope(scope);

		int scopeOffset = lineSelection.x;
		int scopeLength = lineSelection.y;
		int offset = isActive(SearchOptions.FORWARD) ? scopeOffset : scopeOffset + scopeLength;
		extensionTarget.setSelection(offset, 0);
	}
```

Note that the search-scope is set in dependence of whether forward-search is active or not. In my attempts to fix this bug, I set forward-search AFTER setting global search. This means that the search-cursor was set to the end of the end of the scope.